File: agent/comm/layer0/http.go
package layer0

import (
	"encoding/base64"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
)

type HttpClient struct {
	targetUrl string
}

func NewHttpClient(targetUrl string) *HttpClient {
	return &HttpClient{
		targetUrl: targetUrl,
	}
}

func (client *HttpClient) SendMsg(outgoingMsg []byte) ([]byte, error) {
	encodedOutgoingMsg := base64.StdEncoding.EncodeToString(outgoingMsg)
	resp, err := http.PostForm(client.targetUrl, url.Values{"m": {string(encodedOutgoingMsg)}})

	fmt.Printf("Message: %s\n", encodedOutgoingMsg)

	if err != nil {
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("got HTTP status code %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	msg, err := getMsgFromBody(body)
	if err != nil {
		return nil, err
	}

	return msg, nil
}

func getMsgFromBody(body []byte) ([]byte, error) {
	fmt.Printf("%s\n", body)
	re := regexp.MustCompile(`(<!--)([A-Za-z0-9/+=]*|=[^=]|={3,})(-->)`)
	match := re.Find(body)
	encoded := string(match[4 : len(match)-3])
	data, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return nil, err
	}

	return data, nil
}

File: python/malon_agent/malon_agent/comm/http.py
import re
import requests
from typing import Optional
from base64 import b64encode, b64decode

BASE64_MATCH_REGEX = re.compile(b'(<!--)([A-Za-z0-9/+=]*|=[^=]|={3,})(-->)')


def find_encoded_msg(body: bytes) -> Optional[bytes]:
    match = BASE64_MATCH_REGEX.search(body)
    return match[0] if match is not None else None


class HttpClient:
    def __init__(self, url: str):
        self._url = url

    def send_msg(self, msg: bytes) -> bytes:
        response = requests.post(self._url, {'m': b64encode(msg)})
        if not response.ok:
            raise RuntimeError('Got HTTP status code: {}'.format(response.status_code))
        encoded_response_msg = find_encoded_msg(response.content)
        if encoded_response_msg is None:
            raise RuntimeError('No encoded message found in response')

        return b64decode(encoded_response_msg)
